[Project description](https://www.coursera.org/learn/ml-classification/supplement/NtMAS/predicting-sentiment-from-product-reviews)

# Plan
    1. Understand the challange. What questions do I want to answer?
    2. Prepare the data and conduct descriptive analysis.
        A. Load the data
        B. Remove punctuation
        C. Remove n/a fields in the reviews
        D. Exptract sentiments. Ignore neutral reviews with rating 3.
        E. Create a new column that assign positive or negative value for each of the review.
    3. Choose and create a model. Train the model.
        A. Choose the right model.
        B. Split the dataset into training and test.
        C. Create vectors(train, test) using words in the reviews. Use common bug-of-words representation. Choose words from train dataset.
        D. Trein logistic regression model.
    4. Test the model. Interpret the results.
        A. Count all positive weights in the model.
        B. Test the model with simple test data.
        C. Calculate probabilities for sample test data.
        D. Estimate an accuracy of the model.
    5. Answer the questions.

1. The purpose of the project is to predict a sentiment of a product review. The model should predict does a review is positive or negative?

2.A. Load the data.

In [1]:
import pandas as pd
import string as st
import numpy as np
from sklearn.cross_validation import train_test_split

In [2]:
products = pd.read_csv( "./amazon_baby.csv" )
print( products )

                                                     name  \
0                                Planetwise Flannel Wipes   
1                                   Planetwise Wipe Pouch   
2                     Annas Dream Full Quilt with 2 Shams   
3       Stop Pacifier Sucking without tears with Thumb...   
4       Stop Pacifier Sucking without tears with Thumb...   
5       Stop Pacifier Sucking without tears with Thumb...   
6                 A Tale of Baby's Days with Peter Rabbit   
7       Baby Tracker&reg; - Daily Childcare Journal, S...   
8       Baby Tracker&reg; - Daily Childcare Journal, S...   
9       Baby Tracker&reg; - Daily Childcare Journal, S...   
10      Baby Tracker&reg; - Daily Childcare Journal, S...   
11      Baby Tracker&reg; - Daily Childcare Journal, S...   
12      Baby Tracker&reg; - Daily Childcare Journal, S...   
13      Baby Tracker&reg; - Daily Childcare Journal, S...   
14         Nature's Lullabies First Year Sticker Calendar   
15         Nature's Lull

2.B. Remove punctuation from the reviews.

In [3]:
import string as st

def remove_punctuation(text):
    translator = str.maketrans({key: None for key in st.punctuation})
    text = str( text )
    text = text.translate( translator )
    return text

products[ 'review_clean' ] = products[ 'review' ].apply( remove_punctuation )
print( products[ 'review_clean' ] )

0         These flannel wipes are OK but in my opinion n...
1         it came early and was not disappointed i love ...
2         Very soft and comfortable and warmer than it l...
3         This is a product well worth the purchase  I h...
4         All of my kids have cried nonstop when I tried...
5         When the Binky Fairy came to our house we didn...
6         Lovely book its bound tightly so you may not b...
7         Perfect for new parents We were able to keep t...
8         A friend of mine pinned this product on Pinter...
9         This has been an easy way for my nanny to reco...
10        I love this journal and our nanny uses it ever...
11        This book is perfect  Im a first time new mom ...
12        I originally just gave the nanny a pad of pape...
13        I thought keeping a simple handwritten journal...
14        Space for monthly photos info and a lot of use...
15        I bought this calender for myself for my secon...
16        I love this little calender yo

2.C. 

In [4]:
products = products.fillna({'review':''})  # fill in N/A's in the review column

2.D.

In [5]:
products = products[products['rating'] != 3]

products['sentiment'] = products['rating'].apply(lambda rating : +1 if rating > 3 else -1)
print( products[ "sentiment" ] )

1         1
2         1
3         1
4         1
5         1
6         1
7         1
8         1
9         1
10        1
11        1
12        1
14        1
15        1
16        1
17        1
18        1
19        1
20        1
21       -1
22        1
24        1
25        1
26        1
28        1
29        1
30        1
31        1
32        1
33        1
         ..
183499    1
183500    1
183501    1
183502    1
183503   -1
183504    1
183505    1
183506    1
183507    1
183508    1
183509    1
183510    1
183512    1
183513    1
183514    1
183515    1
183517    1
183518    1
183519    1
183520    1
183521    1
183522    1
183523    1
183524    1
183525    1
183526    1
183527    1
183528    1
183529    1
183530    1
Name: sentiment, dtype: int64


3.A. 

I should construct an classifier because I want to classify positive/negative. For the first time let me try logistic regression.

3.B.

In [6]:
train_data, test_data = train_test_split( products, test_size = 0.2)

test_data.reset_index(drop=True, inplace = True)
train_data.reset_index(drop=True, inplace = True)

print( test_data )
print( train_data )

                                                    name  \
0                            Kel-Gar Snug Tub - Elephant   
1      [Award winning] Kidsme Food Feeder (Small size...   
2      Dr. Brown's 3 Pack BPA Free Polypropylene Bott...   
3        Ameda Purely Yours Ultra Transformer AC Adapter   
4      The First Years 6 Count Take &amp; Toss Spill ...   
5                           Bright Starts Starry Teether   
6      Baby Washable Reusable Cloth Pocket Diaper wit...   
7                     OXO Tot Sprout Chair, Orange/Birch   
8      Pumpin' Pal Super Shields, Angled Pumping Flan...   
9                      Baby Buddy: Baby's 1st Toothbrush   
10     Baby Jogger City Mini GT Single Stroller, Shad...   
11                  Baby's First Laptop Computer Playset   
12                    Dr. Brown's Formula Mixing Pitcher   
13     Britax Boulevard Convertible Car Seat, Brownstone   
14     Summer Infant Secure Sight Digital Color Video...   
15                         NoJo Jungle T

3.C.

In [7]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer(token_pattern=r'\b\w+\b') #The simplest pattern
train_matrix = vectorizer.fit_transform(train_data['review_clean'])
test_matrix = vectorizer.transform(test_data['review_clean'])

print( train_matrix )
print( test_matrix )


  (0, 35343)	1
  (0, 59672)	1
  (0, 118052)	1
  (0, 53988)	1
  (0, 119997)	1
  (0, 34966)	1
  (0, 119200)	1
  (0, 107156)	1
  (0, 11979)	1
  (0, 7114)	3
  (0, 70625)	2
  (0, 67669)	1
  (0, 115147)	1
  (0, 81329)	1
  (0, 10281)	4
  (0, 37306)	1
  (0, 108591)	2
  (0, 113903)	1
  (0, 44020)	1
  (0, 113373)	1
  (0, 112675)	1
  (0, 37201)	1
  (0, 118184)	1
  (0, 57773)	1
  (0, 58777)	2
  :	:
  (133400, 22363)	1
  (133400, 82687)	1
  (133400, 86575)	1
  (133400, 41127)	1
  (133400, 60307)	1
  (133400, 101313)	1
  (133400, 88584)	1
  (133400, 41931)	1
  (133400, 71732)	1
  (133400, 59916)	1
  (133400, 26410)	1
  (133400, 69211)	1
  (133400, 62447)	1
  (133400, 93074)	1
  (133400, 24426)	1
  (133400, 114845)	1
  (133400, 86624)	1
  (133400, 28168)	1
  (133400, 92660)	1
  (133400, 10270)	1
  (133400, 90380)	1
  (133400, 86182)	1
  (133400, 78966)	1
  (133400, 30624)	1
  (133400, 45622)	1
  (0, 915)	1
  (0, 1980)	1
  (0, 2450)	1
  (0, 9331)	1
  (0, 10281)	2
  (0, 12387)	1
  (0, 13383)	1
  (0, 24

3.D.

In [8]:
from sklearn import linear_model

sentiment_model = linear_model.LogisticRegression()

sentiment_model.fit( train_matrix, train_data[ 'sentiment' ] )

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [9]:
print( sentiment_model.coef_ )
print( len( sentiment_model.coef_[ 0 ] ) )

[[ -8.79531107e-01   1.79593769e-02   7.36696321e-03 ...,   2.75774122e-05
    5.70562429e-03   5.68510084e-03]]
121350


## 4.A.

In [10]:
number_positive_coefficients = 0
for coefficient in sentiment_model.coef_[ 0 ]:
    if coefficient >= 0:
        number_positive_coefficients += 1
        
print( number_positive_coefficients )

85544


## **Conclusion:**
    Most of the reviews are positive.
    
## 4.B.

In [11]:
sample_test_data = test_data[10:13]
print( sample_test_data )

                                                 name  \
10  Baby Jogger City Mini GT Single Stroller, Shad...   
11               Baby's First Laptop Computer Playset   
12                 Dr. Brown's Formula Mixing Pitcher   

                                               review  rating  \
10  In 2 years, this is my 5th stroller.  I resear...       5   
11  RUBISH! DONT LIKE THIS AT ALL YOU GUYS SHOULD ...       1   
12  My triplets get half formula and half breastmi...       4   

                                         review_clean  sentiment  
10  In 2 years this is my 5th stroller  I research...          1  
11  RUBISH DONT LIKE THIS AT ALL YOU GUYS SHOULD S...         -1  
12  My triplets get half formula and half breastmi...          1  


In [12]:
print( sample_test_data['review'][10] )

In 2 years, this is my 5th stroller.  I researched and watched you tube video's over and over trying to find the perfect stroller.  Each stroller I purchased, I felt it was perfect...that perfect lasted a few short months. I was a big britax fan having the b-ready (felt it was too big and sold it) then purchasing the b-agile.I really like the b-safe infant car seat and hesitated giving up my b-agile because of the snap and go compatibility, however now being compatitible with Britax I decided to sell my b-agile and buy the CM.B-Agile and City Mini GT are very similar but for me there were two big differences1. although looks the same, the CM basket is larger and accessible from the side. Makes a huge difference when trying to get things in and out.2. Adjustable handle barThese two points above have made all the difference between the b-agile and CM. I love the CM and would definitely recommend it to a friend. I also find having the forever tires,  its sturdier than the b-agile and is s

In [13]:
print( sample_test_data['review'][11] )

RUBISH! DONT LIKE THIS AT ALL YOU GUYS SHOULD STOP SELL IT DOSENT WORTH THE MONEY I WOULD NEVER ORDER IT AGAIN


In [14]:
print( sample_test_data['review'][12] )

My triplets get half formula and half breastmilk. This pitcher allows us to make more than one feeding worth of formula. All the formula gets disolved with no clumps.Only downside is that it can squirt out the top when mixing. It says it can do 32 oz, but 28 is the max for nothing coming out the top when you mix it.


In [15]:
sample_test_matrix = vectorizer.transform(sample_test_data['review_clean'])
scores = sentiment_model.decision_function(sample_test_matrix)
print( scores )

[ 8.76207206 -3.75180708  3.24422683]


## Note:
    Based on analysis of the sample_test_data the model works! Review 11 is the most positive and has the highest score.

In [16]:
print( sentiment_model.predict( sample_test_matrix ) )

[ 1 -1  1]


In [17]:
print( sample_test_data[ 'sentiment' ] )

10    1
11   -1
12    1
Name: sentiment, dtype: int64


## Conclusions:
    The results above show that the model got 100% accuracy on sample_test_data.
    
## 4.C.
    In order to calculate probabilities I will use the following formula:
    
$$
P(y_{i}|x_{i},w)=\frac{1}{1-exp(-w^{T}h(x_{i}))}
$$


In [20]:
def probability_logistic_regression( score ):
    return 1 / ( 1 + np.exp( score ) )

for score in scores:
    print( probability_logistic_regression( - score ) )

0.999843464636
0.0229368370723
0.962465106817


## 4.D.

In order to calcualte accuracy I will use the following formula:
$$
accuracy = \frac{\text{# correctly clasified}}{\text{# all reviews}}
$$

In [29]:
correctly_classified = 0
#train_data[ 'sentiment' ]
test_sentiments_predictions = sentiment_model.predict( test_matrix )
#print( len( test_data ) )
#print( len( test_sentiments_predictions ) )

for i in range( len( test_data ) ):
    if test_data.loc[ i, 'sentiment' ] == test_sentiments_predictions[ i ]:
        correctly_classified += 1

print( "%.2f" % ( correctly_classified / len( train_data ) ) )

0.23


## Comment
    The sentiment_model accuracy is quite poor. Just 0.23